crates/forge/bin/cmd/script/build.rs

```diff
@@ -101,109 +97,86 @@ impl ScriptArgs {
-                        }
-                        *extra.dependencies = unique_deps(dependencies);
-                        *extra.contract = contract.clone();
-                        extra.matched = true;
-                        extra.target_id = Some(id.clone());
-                    }
-                } else {
-                    let (path, name) = extra
-                        .target_fname
-                        .rsplit_once(':')
-                        .expect("The target specifier is malformed.");
-                    let path = std::path::Path::new(path);
-                    if path == id.source && name == id.name {
-                        *extra.dependencies = unique_deps(dependencies);
-                        *extra.contract = contract.clone();
-                        extra.matched = true;
-                        extra.target_id = Some(id.clone());
-                    }
+            } else {
+                let (path, name) =
+                    target_fname.rsplit_once(':').expect("The target specifier is malformed.");
+                let path = std::path::Path::new(path);
+                if path == id.source && name == id.name {
+                    target = Some(id);
                 }
+            }
+        }
 
-                if let Ok(tc) = ContractBytecode::from(contract).try_into() {
-                    highlevel_known_contracts.insert(id, tc);
-                }
-
-                Ok(())
-            },
-            project.root(),
-        )?;
-
-        let target = extra_info
-            .target_id
-            .ok_or_else(|| eyre::eyre!("Could not find target contract: {}", target_fname))?;
+        target.ok_or_eyre(format!("Could not find target contract: {}", target_fname))
+    }
 
-        let (new_libraries, predeploy_libraries): (Vec<_>, Vec<_>) =
-            run_dependencies.into_iter().unzip();
+    pub fn link(
+        &self,
+        project: Project,
+        contracts: ArtifactContracts,
+        libraries: Libraries,
+        sender: Address,
+        nonce: u64,
+        target: ArtifactId,
+    ) -> Result<BuildOutput> {
+        let LinkOutput { libs_to_deploy, contracts: linked_contracts, predeployed_libs } =
+            link_with_nonce_or_address(&contracts, &libraries, sender, nonce, &target)?;
 
         // Merge with user provided libraries
-        let mut new_libraries = Libraries::parse(&new_libraries)?;
-        for (file, libraries) in libraries_addresses.libs.into_iter() {
-            new_libraries.libs.entry(file).or_default().extend(libraries)
+        let mut new_libraries = Libraries { libs: BTreeMap::new() };
+        for (id, address) in libs_to_deploy.iter().chain(predeployed_libs.iter()) {
+            new_libraries
+                .libs
+                .entry(id.source.clone())
+                .or_default()
+                .insert(id.name.split('.').next().unwrap().to_owned(), address.to_string());
```

Comment:
When we need to deploy libraries, it might be nice to do it with create2 (if we don't already). Using create2 when a hardcoded create2 deployer is available on the chain seems preferable because then libs can be automatically reused across projects without needing to redeploy (see also: https://github.com/foundry-rs/foundry/issues/4810)

Reply:
this is a nice idea! I was also thinking about it in scope of #5375. I think it should be much easier to implement with less complex linker implementation, but imo it makes more sense as a separate PR as it'd be a new feature